### ISSUE TITLE
Support for distributed layernorm

### ISSUE THREAD

**Issue by johanna-rock-tt** [tenstorrent/tt-metal#8796]
Currently there is no support for layernorm to be distributed over multiple devices.

Implications
- Activations need to be gathered before the layernorm to process the whole input width per device, replicating the layernorm
- Models with large hidden_size (e.g. falcon40b, llama2/3) can't use the interleaved implementation because it runs out of L1 since there's no parallelization over multiple cores fracturing in activation width
- The block sharded version is functional and correct, but runs out of memory for prefill and large S (e.g. for falcon we need to process in 2-4 slices) resulting in a large number of small ops operating on partial activation slices

Proposed Solution: Implementation of a distributed layernorm
- Processes width fractured input activations in parallel
- Computes per-device statistics and reduces them over all devices to compute the overall statistics, and continues the normalization of the local activation fracture
- Reads interleaved inputs and writes outputs to DRAM interleaved; most relevant for prefill model where we need to read/write to/from DRAM anyway
- Support both variants of normalization: layernorm and RMS Norm

**Comment by TT-BrianLiu** [comment]
So far, we don't have any ops that operate across device other than CCL ops which @SeanNijjar is working on. This op seems like something that should be generalized as we start supporting multi-chip ops, so I think it's important we get things right the first time. (Potentially need a wider design review for how we want to handle multi-chip ops.) For example:
- Are there parts of this op that is covered by existing/upcoming CCL ops? @SeanNijjar is working on all reduce with his new infra.
- Should ops like this be implemented as some variant of a CCL op?
- What happens if input/outputs are sharded?

**Comment by cglagovichTT** [comment]
We are thinking of implementing distributed layernorm as 3 separate programs so we can use Sean's AllGather as it exists today. We would invoke it like this:
```python
stats = tt_lib.tensor.layernorm_tophalf(x)
stats = ttnn.all_gather(stats)
out = tt_lib.tensor.layernorm_bottomhalf(x, epsilon, starts, weight, bias)
```

From talking with Sean, infra to support CCL within ops isn't mature so this is probably the fastest we can bring up this functionality.

**Comment by cglagovichTT** [comment]
We are planning on restricting it to interleaved for two reasons
1. We want to AllGather to produce a tensor of gathered statistics which each core (parallelizing on sequence length, so we have no inter-core comms) can pick up and reduce. Sharded AllGather currently only supports width-sharded, single-tile high tensors, so there's no way to AllGather sharded and assign output tiles to the correct worker core
2. The purpose of distributed layernorm is to push the AllGather down to the Matmul in the MLP, so we can use AllGather/Matmul overlapping. This optimization helps the most for prefill mode, where we expect inputs to be interleaved.

Thinking about it, we could support sharded input/output but I think the stats would have to be interleaved DRAM. But I think models would have to explicitly shard and then interleave input/output to use this, so we would pay non-overlapped reads/writes anyways.

**Comment by TT-BrianLiu** [comment]
> We are thinking of implementing distributed layernorm as 3 separate programs so we can use Sean's AllGather as it exists today. We would invoke it like this:
> 
> ```python
> stats = tt_lib.tensor.layernorm_tophalf(x)
> stats = ttnn.all_gather(stats)
> out = tt_lib.tensor.layernorm_bottomhalf(x, epsilon, starts, weight, bias)
> ```
> 
> From talking with Sean, infra to support CCL within ops isn't mature so this is probably the fastest we can bring up this functionality.

I see. This makes sense. Two things:
- Is `layernorm_tophalf` and `layernorm_bottomhalf` very similar in kernels or completely separate? 
- This all gather + reduce in `layernorm_bottomhalf` sounds like a reduce-scatter CCL op. @SeanNijjar Is this something we will be implementing later?

**Comment by cglagovichTT** [comment]
1. tophalf and bottomhalf are relatively different. tophalf only computes local `sum(x)` and `sum(x**2)`. bottom half takes the gathered stats and computes the rest of the layernorm

**Comment by SeanNijjar** [comment]
> This all gather + reduce in layernorm_bottomhalf sounds like a reduce-scatter CCL op. @SeanNijjar Is this something we will be implementing later?
@TT-BrianLiu - yes this is the plan eventually, but we shouldn't require it to bringup this functionality. The plan is, similar to our op list, to provide all the baseline CCL "primitives": all-gather, all-reduce, reduce-scatter, all-scatter, bcast, send, etc.

Beyond that, I'm hoping to provide a bit more of a programming model (sort of) that should a) make building these ops simpler, b) empower developers like Colman and Johanna to experiment with custom schemes like this without having to depend on me or have a potentially compromised implementation that relies on these bigger ops

**Comment by cglagovichTT** [comment]
@kevinmiTT11 found what might be a bug in distributed layernorm - since pre-allgather rmsnorm produces and output which is `1 x 1 x R x 1[32]` for each device, the allgather afterward produces an output that is `1 x 1 x R x 4[32]` for 4 devices - this messes up the computation in post-allgather rmsnorm, since it expects 4 devices to produce 4 tile columns of stats. @kevinmiTT11 has a workaround which reshapes the output of pre-allgather rmsnorm.

**Comment by SeanNijjar** [comment]
@johanna-rock-tt no this isn't the same issue. The issue I was talking to you about was a separate issue on branch only (was since resolved and merged).

However, @cglagovichTT's description is different from how I understood your offline description. My understanding was that all-gather should be getting as input something like

`[1,1,R,1[32] x ring_size`

with output (per chip)

`[1, 1, R, num_chips x 1[32]]` - in other words, num_chips worth of tiles on width, each having a column of valid data.

**Comment by SeanNijjar** [comment]
@cglagovichTT I did recently add the new, second,  bidirectional mode for all-gather. Fundamentally there should have been no changes. The all gather and distributed layernorm tests are all passing. It should be impossible for the allgather to emit `1, 1, R, num_chips[32]` and that definitely would have been caught by the tests 🤔

**Comment by cglagovichTT** [comment]
Let me dig deeper into my model code... I must be missing something then

**Comment by cglagovichTT** [comment]
Support is in main, closing
